Remove assert around self.pending_capacity.is_empty() (#224)

DarrenTsung opened the pull request:
This assert does not hold as many streams can be pushed into
pending_capacity during a call to send_data(). See issue #224
for more discussion and sign-off.

DarrenTsung commented:
Oops, got to remove some unused methods.

DarrenTsung commented:
Ok, seems good to me. Tests pass with `cargo test --features unstable`.

DarrenTsung commented:
Currently looking at writing a test for demonstrating the panic..

DarrenTsung commented:
Ok, added a failing test. Basically given some set of queued requests that are larger than the default window size, send_data should not panic when assigning more capacity to a stream.

Pretty much all of the new data that the mock receives I have no idea how to interpret, maybe you guys could help me out here so the test doesn't look like magic? The base of the test is taken from the test `single_stream_send_body_greater_than_default_window()` which is a couple lines below.

https://github.com/DarrenTsung/h2/commit/4fd36052a03a769e31f5e533bc5adecd72f1c1e6#diff-726773c7c7a823eac63a932c9c0f117bR83

DarrenTsung commented:
Can anyone take a look the test added in this commit: https://github.com/carllerche/h2/pull/225/commits/4fd36052a03a769e31f5e533bc5adecd72f1c1e6?

I think it demonstrates the failing case correctly, but I'm not sure why some of the data is expected / how to document the test correctly.

carllerche commented:
It looks like the test is hitting hte assertion. I'm not sure what you mean by:

> but I'm not sure why some of the data is expected / how to document the test correctly.

Also, is there a reason you wrote the raw frame bytes instead of using the [helpers](https://github.com/carllerche/h2/blob/master/tests/ping_pong.rs#L46-L53)?

DarrenTsung commented:
Sorry for the delay, yes, the test is hitting the assertion which is what I wanted to make sure was covered.

Basically, nearly all of the expected data for the test is taken from another test in the file `single_stream_send_body_greater_than_default_window()` which is a [couple lines below](https://github.com/carllerche/h2/blob/4fd36052a03a769e31f5e533bc5adecd72f1c1e6/tests/prioritization.rs#L251). I added [a couple sections in the expected data with raw frame bytes](https://github.com/carllerche/h2/blob/4fd36052a03a769e31f5e533bc5adecd72f1c1e6/tests/prioritization.rs#L83) in response to the unexpected (new) data from the new `send_request` calls.

Everything works and the test fails as expected, which is great. The only thing is I'm not sure exactly what the new data receiving means. Some of it looks like an ACK for the `send_request` calls I added (my guess), but some of it does not look like that at all. I would love to use the helpers for the data to make it more read-able / comprehensible, just not sure which helpers to use.

Any guidance on interpreting the new data, either through comments or using a helper instead, would be very appreciated!

DarrenTsung commented:
Can anyone take a look at the test above? If it seems fine I can remove the assert, have the test pass, and the PR should be ready for merge (unless any changes are requested).

seanmonstar commented:
I admit that looking at the test and seeing raw bytes makes it a lot harder to review and then also be able to debug in the future if we every accidentally regress. For instance, I'm not actually sure what part is unexpected, nor what the `// (???)` mean :D

What do think of tests that are written in this style instead? https://github.com/carllerche/h2/blob/master/tests/flow_control.rs#L1162-L1204

DarrenTsung commented:
Thanks for pointing out that example @seanmonstar! I was able to get rid of the "hand-wave-y" raw data frames and pushed a commit that fixed the test by removing the debug assertion again.
